streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -190,11 +190,42 @@ public void print(Serde<K> keySerde, Serde<V> valSerde) {
 
     @Override
     public void print(Serde<K> keySerde, Serde<V> valSerde, String streamName) {
+        print(keySerde, valSerde, streamName, null);
+    }
+
+    @Override
+    public void print(final Serde<K> keySerde, final Serde<V> valSerde, String streamName, final KeyValueMapper<K, V, String> mapper){
         String name = topology.newName(PRINTING_NAME);
         streamName = (streamName == null) ? this.name : streamName;
-        topology.addProcessor(name, new KeyValuePrinter<>(keySerde, valSerde, streamName), this.name);
+        topology.addProcessor(name, new KStreamPeek<>(printAction(System.out, keySerde, valSerde, streamName, mapper)), this.name);
     }
 
+    private static <K, V> ForeachAction<K, V> printAction(final PrintStream printStream, final Serde<?> keySerde, final Serde<?> valueSerde, final String streamName, final KeyValueMapper<K, V, String> mapper) {
+        return new ForeachAction<K, V>() {
+            @Override
+            public void apply(final K key, final V value) {
+                K keyToPrint = (K) maybeDeserialize(key, keySerde.deserializer());
+                V valueToPrint = (V) maybeDeserialize(value, valueSerde.deserializer());
+                if(mapper == null) {
+                    printStream.println("[" + streamName + "]: " + keyToPrint + " , " + valueToPrint);
+                } else {
+                    printStream.println("[" + streamName + "]: " + mapper.apply(keyToPrint, valueToPrint));
+                }
+            }
+
+            private Object maybeDeserialize(Object receivedElement, Deserializer<?> deserializer) {
+                if (receivedElement == null) {
+                    return null;
+                }
+
+                if (receivedElement instanceof byte[]) {
+                    return deserializer.deserialize("Topic", (byte[]) receivedElement);
```

Comment:
Okay, this actually makes the demanded changes a bit tricky. I see basically two ways of tackling this. For both approaches I created a PR and suggest we continue discussion over there: 

1. Topic name is not that important for now, which would allow to use `peek()` very nicely: https://github.com/apache/kafka/pull/2703
2. We maintain `KeyValuePrinter`but refactor such that it extends `KStreamPeek` and uses `ForeachAction`: https://github.com/apache/kafka/pull/2704

I am much more convinced with (1) -- but I'm unsure how important the topic name during the deserialization process actually is.